### PR TITLE
OCPBUGS-8938: OCPBUGS-15202: MCO-555: kube-rbac-proxy addition

### PR DESCRIPTION
**Pull request by cdoern**
**- What I did**

- I replaced all oauth-proxy containers in the daemon and the controller with kube-rbac-proxy containers, mimicking the work done in MCO-407.

**- How to verify it**

Retrieve metrics as usual, they should work without any hiccups using this new image. You can go to the console to easily retrieve metrics. 

**- Description for the changelog**

metrics backend now depends on kube-rbac-proxy instead of oauth-proxy.

### CODE REVIEW

File: manifests/machineconfigcontroller/deployment.yaml [hunk 1]
```diff
@@ -27,31 +27,26 @@ spec:
             cpu: 20m
             memory: 50Mi
         terminationMessagePolicy: FallbackToLogsOnError
-      - name: oauth-proxy
```
Comment (yuqi-zhang) [inline manifests/machineconfigcontroller/deployment.yaml:30]: Looks like there are some unit testing failures, e.g. `TestRenderAsset` is still explicitly referencing OauthProxy

File: manifests/machineconfigcontroller/deployment.yaml [hunk 2]
```diff
@@ -27,31 +27,26 @@ spec:
...
+        - --logtostderr=true
+        - --tls-cert-file=/etc/tls/private/tls.crt
+        - --tls-private-key-file=/etc/tls/private/tls.key
         resources:
           requests:
             cpu: 20m
             memory: 50Mi
         volumeMounts:
-        - mountPath: /etc/tls/private
-          name: proxy-tls
```
Comment (jkyros) [inline manifests/machineconfigcontroller/deployment.yaml:52]: I think you still need to mount the the TLS certificate here (you left it in for the daemon, assuming this was an accident)  

I'll leave the "should these all have their own certificate, or is it cool that they're sharing the same one" question for a later date. :smile:

**Comment by yuqi-zhang** [comment]
I tried to take a look at the e2e-gcp-op failures. The recent failures are very consistent on this PR, e.g. https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/3663/pull-ci-openshift-machine-config-operator-master-e2e-gcp-op/1649037169399959552/artifacts/e2e-gcp-op/test/build-log.txt

Indicates that the expectation is that the cert rotation happened, but it did not in actuality. Looking into it a bit deeper, what seems to happen is this:

I brought up a cluster with this PR, and then ran the test manually, everything passed fine. When the cert rotation was forced, it generated one new rendered-config. Rendered-config-A had 7 certs, rendered-config-B had 9 certs (the rotation regens 2), so that is expected, but very odd

Looking at the latest run with this PR, the diff indicates that the MCD has yet to write the latest version. This is because there are two MCs generated:
1. db4b73411bd5d29d16d289bfb1f78589 -> this contains 1/2 new certs (bringing us up to 8), but ALSO had the diff of:
```
<       disks:
<       - device: /one
...
<       - contents:
<           source: data:,test
<         mode: 420
<         overwrite: true
<         path: /etc/mytestconfs
<         user:
<           name: root
```

so it generated a halfway MC. Then 5 seconds later, the MC with both cert changes, bringing up to 9, and the last one was the missing one.

So this is a very weird scenario that I am not sure why this particular test hits. e.g. in a successful run, we don't hit it that close: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/3441/pull-ci-openshift-machine-config-operator-master-e2e-gcp-op/1649034037110509568/artifacts/e2e-gcp-op/gather-extra/artifacts/oc_cmds/machineconfigs

https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/3663/pull-ci-openshift-machine-config-operator-master-e2e-gcp-op/1648450147098562560/artifacts/e2e-gcp-op/gather-extra/artifacts/oc_cmds/machineconfigs/ has the same issue (from this PR), https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/3663/pull-ci-openshift-machine-config-operator-master-e2e-gcp-op/1648511723532980224/artifacts/e2e-gcp-op/gather-extra/artifacts/oc_cmds/machineconfigs is a bit further away, so I am not sure why that also exhibited the issue.

So in the test, we first: rotate the token: https://github.com/openshift/machine-config-operator/blob/master/test/e2e/mcd_test.go#L946

Then we get the incorrect bundle (in theory):
https://github.com/openshift/machine-config-operator/blob/master/test/e2e/mcd_test.go#L952

but this cert is the latest: https://github.com/openshift/machine-config-operator/blob/master/test/e2e/mcd_test.go#L957

but probably at this point this controllerconfig: https://github.com/openshift/machine-config-operator/blob/master/test/e2e/mcd_test.go#L963 has yet to update

so this wait: https://github.com/openshift/machine-config-operator/blob/master/test/e2e/mcd_test.go#L965 and the subsequent read: https://github.com/openshift/machine-config-operator/blob/master/test/e2e/mcd_test.go#L968

Are both wrong? I think that's the most likely scenario.

I... can't really pinpoint why this sequence of events is happening though, since I can't replicate it off a cluster on this PR, and saying "well this just got very unlucky 3 times in a row" seems... wrong?

TLDR: our 5-second rendering delay is messing with this in conjunction with the fact that the recent tests somehow generated half-ready MachineConfigs at the perfect timing. The MCD test can definitely be improved, but I wonder what is causing this weird effect

For now though, can you add a sleep of 10 seconds at this line and see if that passes? I just wanted to check if the final result is correct: https://github.com/openshift/machine-config-operator/blob/master/test/e2e/mcd_test.go#L955

**Comment by yuqi-zhang** [comment]
also for just one more data point:

/test e2e-gcp-op

File: lib/resourcemerge/core.go [hunk 3]
```diff
@@ -90,7 +90,8 @@ func ensurePodSpec(modified *bool, existing *corev1.PodSpec, required corev1.Pod
 	setInt32Ptr(modified, &existing.Priority, required.Priority)
 }
 
-func ensureContainer(modified *bool, existing *corev1.Container, required corev1.Container) {
+func ensureContainer(modified *bool, existing *corev1.Container, required corev1.Container) bool {
+	overlap := false
 	resourcemerge.SetStringIfSet(modified, &existing.Name, required.Name)
```
Comment (yuqi-zhang) [inline lib/resourcemerge/core.go:95]: The latest tests apparently panicked on this line, looks like the... existing container didn't exist?

File: lib/resourcemerge/core.go [hunk 4]
```diff
@@ -39,13 +52,25 @@ func ensurePodSpec(modified *bool, existing *corev1.PodSpec, required corev1.Pod
...
+	// so we need a way to detect if containers with different names are going to collide.
 	for _, required := range required.Containers {
 		var existingCurr *corev1.Container
 		for j, curr := range existing.Containers {
 			if curr.Name == required.Name {
 				existingCurr = &existing.Containers[j]
 				break
+			} else if ensureMinimalOverlap(&curr, &required) {
+				existingCurr = &existing.Containers[j]
+				break
```
Comment (jkyros) [inline lib/resourcemerge/core.go:73]: This is what is was generating before, btw, I had it dump it out: https://gist.github.com/jkyros/b23fecbff849fb391fddffa1fd36d3cf so yep, they were both in there and causing a conflict. 

If I understand this right, you're telling it that the `kube-rbac-proxy` container and the `oauth-proxy` pod are equivalent -- they "match" because they have a port conflict instead of because of their names -- which works in this specific case. 

In the general case though...where they didn't conflict, e.g. we just have a new manifest that adds a new container and removes an old container, we'd get the new container, but we'd still end up stuck with the old container, wouldn't we? (nothing ever removes old containers from the pod spec, does it?)

Comment (cdoern) [reply lib/resourcemerge/core.go:73]: the `ensureMinimalOverLap` block only executes if true (if the ports overlap, but we can add more cases). Let's say we have two containers, with completely different configs, one is old one is new. We would never enter the `else if` since the configs are different

I think you may be right that we never remove old containers... that seems a little problematic come to think of it. It seems that we merge where we can and now avoid overlap but if you have `randomCTR123` and it just so happens to not have any overlap with any new containers you are adding, it will just slip under the radar...

Comment (jkyros) [reply lib/resourcemerge/core.go:73]: >the ensureMinimalOverLap block only executes if true (if the ports overlap, but we can add more cases). Let's say we have two containers, with completely different configs, one is old one is new. We would never enter the else if since the configs are different

I just thought that was a weird corner that if two containers that were otherwise completely different specified the same port resourcemerge would merge them. The recipient of that future surprise would be...well...surprised :smile: 

> I think you may be right that we never remove old containers... that seems a little problematic come to think of it. It seems that we merge where we can and now avoid overlap but if you have randomCTR123 and it just so happens to not have any overlap with any new containers you are adding, it will just slip under the radar...

Yeah, I was thinking "If we cleaned up the existing ones properly or just overwrote the whole container section, the collision wouldn't have ever happened".  

Anyway, all we _really_ need to do is just not leave a minefield in the short term until we switch to the actual `library-go` (instead of our fork) and use[ its functions ](https://github.com/openshift/library-go/blob/add9892da47ef737135dae6a58e33127fd79ba70/pkg/operator/resource/resourceapply/apps.go#L181)-- which only really worry about the generation and merging the metadata and just overwrite everything else. 

(And if now you are like "then couldn't we just do something like upstream `library-go` does for now and overwrite existing.Containers with required.Containers if the lists are different", I think the answer is yes)

**Comment by cdoern** [comment]
/retest-required

really weird timeouts here cc/ @yuqi-zhang @cheesesashimi

**Comment by cdoern** [comment]
trying a slightly different structure where we drop any non mentioned containers on update rather then looking for specific overlap.

**Comment by cdoern** [comment]
@jkyros does this version look a little better for excluding old containers on resourcemerge?

File: lib/resourcemerge/core.go [hunk 5]
```diff
@@ -50,11 +63,12 @@ func ensurePodSpec(modified *bool, existing *corev1.PodSpec, required corev1.Pod
 		}
 		if existingCurr == nil {
 			*modified = true
-			existing.Containers = append(existing.Containers, corev1.Container{})
-			existingCurr = &existing.Containers[len(existing.Containers)-1]
+			existingCurr = &corev1.Container{}
```
Comment (jkyros) [inline lib/resourcemerge/core.go:66]: So if I'm interpreting this right, this is "if there's one in the old list with the same name as the new list, replace it with an empty object so the new object's fields always "win" the merge" ? 

The new ones always make it in, the old ones that match get blanked to be overwritten, and the old ones that don't match never make it into the list. 

That seems like we eventually get the outcome we want, but you think this is buying us anything over just  replacing the list if it's different, e.g.: 
```
if !equality.Semantic.DeepEqual(required.Containers, *existing.Containers) {
		*existing.Containers = required.Containers
		*modified = true
}
```

Comment (cdoern) [reply lib/resourcemerge/core.go:66]: I think this may work @jkyros. I am trying to think of scenarios where ensureContainer doesn't just take required (the new ctr) but i can't seem to find one

the way we do things now in this PR:

OK, for each required container, if an existing ctr exists with the same name, compare them and merge the entities but really just take everything from required and get rid of existing.

the deep equal way:

OK, if the new set of containers differs from the old one, take the entire new set.

Seems like the same outcome to me, lets try it.

File: pkg/operator/sync.go [hunk 6]
```diff
@@ -685,15 +726,20 @@ func (optr *Operator) syncMachineConfigDaemon(config *renderConfig) error {
...
 
 	// Only generate a new proxy cookie secret if the secret does not exist or if it has been deleted.
-	_, err := optr.kubeClient.CoreV1().Secrets(config.TargetNamespace).Get(context.TODO(), "cookie-secret", metav1.GetOptions{})
-	if apierrors.IsNotFound(err) {
-		paths.secrets = []string{"manifests/machineconfigdaemon/cookie-secret.yaml"}
-	} else if err != nil {
-		return err
-	}
+	/*
+		_, err := optr.kubeClient.CoreV1().Secrets(config.TargetNamespace).Get(context.TODO(), "cookie-secret", metav1.GetOptions{})
```
Comment (cheesesashimi) [inline pkg/operator/sync.go:736]: Does this still need to be commented out? If so, we should delete it completely to avoid confusion.

File: pkg/operator/sync.go [hunk 7]
```diff
@@ -560,6 +578,18 @@ func (optr *Operator) applyManifests(config *renderConfig, paths manifestPaths)
 		}
 	}
 
+	for _, path := range paths.configMaps {
```
Comment (cheesesashimi) [inline pkg/operator/sync.go:581]: This could be useful for the on-cluster build work in the future! cc: @dkhater-redhat.

**Review by cheesesashimi — COMMENTED** [review]
Overall, this looks good. I just have one question about a section of commented-out code.

**Review by yuqi-zhang — APPROVED** [review]
/lgtm

Seems like a clean replacement from oauth->kube-rbac, lgtm!

Given that there are three associated cards, adding a

/hold

Looking at the associated bugs, I don't think QE verification is directly needed for either of them. @sergiordlr would you like to pre-merge test the kube-rbac-proxy? if not, I think we can unhold and merge, and follow up via nightly builds to see if previous issues still exist

**Comment by rioliu-rh** [comment]
per-merge test:
build cluster with this PR
check whether container `kube-rbac-proxy` in mcc/mcd pod: pass
```console
# mcc pod
oc get pod/machine-config-controller-8777fc4c6-zkw2l -n openshift-machine-config-operator -o yaml | yq -y '.status.containerStatuses'
- containerID: cri-o://f9e429db18e9171c783e35c7c6c3e2ca682378ebc0f2f0be68ea391655e0aeab
  image: registry.build05.ci.openshift.org/ci-ln-xkxzr22/stable@sha256:b5e574c5b2fdd0a90d899793e2ce97792dc2d3e3fbf934a107139b1b4f2732a7
  imageID: registry.build05.ci.openshift.org/ci-ln-xkxzr22/stable@sha256:b5e574c5b2fdd0a90d899793e2ce97792dc2d3e3fbf934a107139b1b4f2732a7
  lastState: {}
  name: kube-rbac-proxy
  ready: true
  restartCount: 0
  started: true
  state:
    running:
      startedAt: '2023-08-11T07:03:46Z'
...

# mcd pod
oc get pod/machine-config-daemon-8n9zq -n openshift-machine-config-operator -o yaml |  yq -y '.status.containerStatuses'
- containerID: cri-o://5969e1c0ad6d1a55e1ee8e7bcbec8fe89c1c7b3fd394a9d5c8c28d63b317fb63
  image: registry.build05.ci.openshift.org/ci-ln-xkxzr22/stable@sha256:b5e574c5b2fdd0a90d899793e2ce97792dc2d3e3fbf934a107139b1b4f2732a7
  imageID: registry.build05.ci.openshift.org/ci-ln-xkxzr22/stable@sha256:b5e574c5b2fdd0a90d899793e2ce97792dc2d3e3fbf934a107139b1b4f2732a7
  lastState: {}
  name: kube-rbac-proxy
  ready: true
  restartCount: 0
  started: true
  state:
    running:
      startedAt: '2023-08-11T07:11:47Z'
...
```
check enabled cipher definition `kube-rbac-proxy` containers, no weak ciphers mentioned in the bug [OCPBUGS-15202](https://issues.redhat.com/browse/OCPBUGS-15202)
```console
 oc get pod/machine-config-controller-8777fc4c6-zkw2l -n openshift-machine-config-operator -o yaml | yq -y '.spec.containers[]|select(.name=="kube-rbac-proxy")' | grep tls-cipher-suites
  - --tls-cipher-suites=TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305

oc get pod/machine-config-daemon-8n9zq -n openshift-machine-config-operator -o yaml | yq -y '.spec.containers[]|select(.name=="kube-rbac-proxy")' | grep tls-cipher-suites
  - --tls-cipher-suites=TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305
```
connect to `kube-rbac-proxy` service (`secure-listen-address=0.0.0.0:9001`) to verify the ciphers
```console

sh-5.1# ./testssl.sh localhost:9001

###########################################################
    testssl.sh       3.2rc2 from https://testssl.sh/dev/
    (c888475 2023-07-31 17:15:15)

      This program is free software. Distribution and
             modification under GPLv2 permitted.
      USAGE w/o ANY WARRANTY. USE IT AT YOUR OWN RISK!

       Please file bugs @ https://testssl.sh/bugs/

###########################################################

 Using "OpenSSL 1.0.2-bad (1.0.2k-dev)" [~179 ciphers]
 on ip-10-0-114-92:./bin/openssl.Linux.x86_64
 (built: "Sep  1 14:03:44 2022", platform: "linux-x86_64")


 Start 2023-08-11 08:27:59                -->> 127.0.0.1:9001 (localhost) <<--

 rDNS (127.0.0.1):       localhost.
 Service detected:       HTTP


 Testing protocols via sockets except NPN+ALPN

 SSLv2      not offered (OK)
 SSLv3      not offered (OK)
 TLS 1      not offered
 TLS 1.1    not offered
 TLS 1.2    offered (OK)
 TLS 1.3    offered (OK): final
 NPN/SPDY   not offered
 ALPN/HTTP2 h2, http/1.1 (offered)

 Testing cipher categories

 NULL ciphers (no encryption)                      not offered (OK)
 Anonymous NULL Ciphers (no authentication)        not offered (OK)
 Export ciphers (w/o ADH+NULL)                     not offered (OK)
 LOW: 64 Bit + DES, RC[2,4], MD5 (w/o export)      not offered (OK)
 Triple DES Ciphers / IDEA                         not offered
 Obsoleted CBC ciphers (AES, ARIA etc.)            not offered
 Strong encryption (AEAD ciphers) with no FS       not offered
 Forward Secrecy strong encryption (AEAD ciphers)  offered (OK)


 Testing server's cipher preferences

Hexcode  Cipher Suite Name (OpenSSL)       KeyExch.   Encryption  Bits     Cipher Suite Name (IANA/RFC)
-----------------------------------------------------------------------------------------------------------------------------
SSLv2
 -
SSLv3
 -
TLSv1
 -
TLSv1.1
 -
TLSv1.2 (server order -- server prioritizes ChaCha ciphers when preferred by clients)
 xc02f   ECDHE-RSA-AES128-GCM-SHA256       ECDH 521   AESGCM      128      TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
 xc030   ECDHE-RSA-AES256-GCM-SHA384       ECDH 521   AESGCM      256      TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
 xcca8   ECDHE-RSA-CHACHA20-POLY1305       ECDH 521   ChaCha20    256      TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
TLSv1.3 (no server order, thus listed by strength)
 x1302   TLS_AES_256_GCM_SHA384            ECDH 253   AESGCM      256      TLS_AES_256_GCM_SHA384
 x1303   TLS_CHACHA20_POLY1305_SHA256      ECDH 253   ChaCha20    256      TLS_CHACHA20_POLY1305_SHA256
 x1301   TLS_AES_128_GCM_SHA256            ECDH 253   AESGCM      128      TLS_AES_128_GCM_SHA256

 Has server cipher order?     yes (OK) -- only for < TLS 1.3


 Testing robust forward secrecy (FS) -- omitting Null Authentication/Encryption, 3DES, RC4

 FS is offered (OK)           TLS_AES_256_GCM_SHA384 TLS_CHACHA20_POLY1305_SHA256 ECDHE-RSA-AES256-GCM-SHA384 ECDHE-RSA-CHACHA20-POLY1305 TLS_AES_128_GCM_SHA256 ECDHE-RSA-AES128-GCM-SHA256
 Elliptic curves offered:     prime256v1 secp384r1 secp521r1 X25519
 TLS 1.2 sig_algs offered:    RSA+SHA1 RSA+SHA256 RSA+SHA384 RSA+SHA512 RSA-PSS-RSAE+SHA256 RSA-PSS-RSAE+SHA384 RSA-PSS-RSAE+SHA512
 TLS 1.3 sig_algs offered:    RSA-PSS-RSAE+SHA256 RSA-PSS-RSAE+SHA384 RSA-PSS-RSAE+SHA512

 Testing server defaults (Server Hello)

 TLS extensions (standard)    "session ticket/#35" "renegotiation info/#65281" "EC point formats/#11" "supported versions/#43" "key share/#51" "application layer protocol negotiation/#16"
 Session Ticket RFC 5077 hint no -- no lifetime advertised
 SSL Session ID support       yes
 Session Resumption           Tickets no, ID: no
 TLS clock skew               Random values, no fingerprinting possible
 Certificate Compression      none
 Client Authentication        optional
 CA List for Client Auth      empty
 Signature Algorithm          SHA256 with RSA
 Server key size              RSA 2048 bits (exponent is 65537)
 Server key usage             Digital Signature, Key Encipherment
 Server extended key usage    TLS Web Server Authentication
 Serial                       282561969A094C6B (OK: length 8)
 Fingerprints                 SHA1 01589FD49B38FC43EC51A95D143F97A7EA8EBCAA
                              SHA256 6C0CB4884FABACA4D2937B4D4F7F320894A61384C13ED0B4FB2C27AE0E1A865B
 Common Name (CN)             machine-config-daemon.openshift-machine-config-operator.svc
 subjectAltName (SAN)         machine-config-daemon.openshift-machine-config-operator.svc machine-config-daemon.openshift-machine-config-operator.svc.cluster.local
 Trust (hostname)             certificate does not match supplied URI (same w/o SNI)
 Chain of trust               NOT ok (self signed CA in chain)
 EV cert (experimental)       no
 Certificate Validity (UTC)   729 >= 60 days (2023-08-11 07:03 --> 2025-08-10 07:03)
                              > 398 days issued after 2020/09/01 is too long
 ETS/"eTLS", visibility info  not present
 Certificate Revocation List  --
 OCSP URI                     --
                              NOT ok -- neither CRL nor OCSP URI provided
 OCSP stapling                not offered
 OCSP must staple extension   --
 DNS CAA RR (experimental)    not offered
 Certificate Transparency     --
 Certificates provided        2
 Issuer                       openshift-service-serving-signer@1691737396
 Intermediate cert validity   #1: ok > 40 days (2025-10-09 07:03). openshift-service-serving-signer@1691737396 <-- openshift-service-serving-signer@1691737396
 Intermediate Bad OCSP (exp.) Ok


 Testing HTTP header response @ "/"

 HTTP Status Code             401 Unauthorized
 HTTP clock skew              0 sec from localtime
 Strict Transport Security    not offered
 Public Key Pinning           --
 Server banner                (no "Server" line in header, interesting!)
 Application banner           --
 Cookie(s)                    (none issued at "/") -- maybe better try target URL of 30x
 Security headers             X-Content-Type-Options: nosniff
 Reverse Proxy banner         --


 Testing vulnerabilities

 Heartbleed (CVE-2014-0160)                not vulnerable (OK), no heartbeat extension
 CCS (CVE-2014-0224)                       not vulnerable (OK)
 Ticketbleed (CVE-2016-9244), experiment.  not vulnerable (OK)
 ROBOT                                     Server does not support any cipher suites that use RSA key transport
 Secure Renegotiation (RFC 5746)           supported (OK)
 Secure Client-Initiated Renegotiation     not vulnerable (OK)
 CRIME, TLS (CVE-2012-4929)                not vulnerable (OK)
 BREACH (CVE-2013-3587)                    no gzip/deflate/compress/br HTTP compression (OK)  - only supplied "/" tested
 POODLE, SSL (CVE-2014-3566)               not vulnerable (OK), no SSLv3 support
 TLS_FALLBACK_SCSV (RFC 7507)              No fallback possible (OK), no protocol below TLS 1.2 offered
 SWEET32 (CVE-2016-2183, CVE-2016-6329)    not vulnerable (OK)
 FREAK (CVE-2015-0204)                     not vulnerable (OK)
 DROWN (CVE-2016-0800, CVE-2016-0703)      not vulnerable on this host and port (OK)
                                           make sure you don't use this certificate elsewhere with SSLv2 enabled services, see
                                           https://search.censys.io/search?resource=hosts&virtual_hosts=INCLUDE&q=6C0CB4884FABACA4D2937B4D4F7F320894A61384C13ED0B4FB2C27AE0E1A865B
 LOGJAM (CVE-2015-4000), experimental      not vulnerable (OK): no DH EXPORT ciphers, no DH key detected with <= TLS 1.2
 BEAST (CVE-2011-3389)                     not vulnerable (OK), no SSL3 or TLS1
 LUCKY13 (CVE-2013-0169), experimental     not vulnerable (OK)
 Winshock (CVE-2014-6321), experimental    not vulnerable (OK)
 RC4 (CVE-2013-2566, CVE-2015-2808)        no RC4 ciphers detected (OK)


 Running client simulations (HTTP) via sockets

 Browser                      Protocol  Cipher Suite Name (OpenSSL)       Forward Secrecy
------------------------------------------------------------------------------------------------
 Android 6.0                  TLSv1.2   ECDHE-RSA-AES128-GCM-SHA256       256 bit ECDH (P-256)
 Android 7.0 (native)         TLSv1.2   ECDHE-RSA-AES128-GCM-SHA256       256 bit ECDH (P-256)
 Android 8.1 (native)         TLSv1.2   ECDHE-RSA-AES128-GCM-SHA256       253 bit ECDH (X25519)
 Android 9.0 (native)         TLSv1.3   TLS_AES_128_GCM_SHA256            253 bit ECDH (X25519)
 Android 10.0 (native)        TLSv1.3   TLS_AES_128_GCM_SHA256            253 bit ECDH (X25519)
 Android 11 (native)          TLSv1.3   TLS_AES_128_GCM_SHA256            253 bit ECDH (X25519)
 Android 12 (native)          TLSv1.3   TLS_AES_128_GCM_SHA256            253 bit ECDH (X25519)
 Chrome 79 (Win 10)           TLSv1.3   TLS_AES_128_GCM_SHA256            253 bit ECDH (X25519)
 Chrome 101 (Win 10)          TLSv1.3   TLS_AES_128_GCM_SHA256            253 bit ECDH (X25519)
 Firefox 66 (Win 8.1/10)      TLSv1.3   TLS_AES_128_GCM_SHA256            253 bit ECDH (X25519)
 Firefox 100 (Win 10)         TLSv1.3   TLS_AES_128_GCM_SHA256            253 bit ECDH (X25519)
 IE 6 XP                      No connection
 IE 8 Win 7                   No connection
 IE 8 XP                      No connection
 IE 11 Win 7                  No connection
 IE 11 Win 8.1                No connection
 IE 11 Win Phone 8.1          No connection
 IE 11 Win 10                 TLSv1.2   ECDHE-RSA-AES128-GCM-SHA256       256 bit ECDH (P-256)
 Edge 15 Win 10               TLSv1.2   ECDHE-RSA-AES128-GCM-SHA256       253 bit ECDH (X25519)
 Edge 101 Win 10 21H2         TLSv1.3   TLS_AES_128_GCM_SHA256            253 bit ECDH (X25519)
 Safari 12.1 (iOS 12.2)       TLSv1.3   TLS_CHACHA20_POLY1305_SHA256      253 bit ECDH (X25519)
 Safari 13.0 (macOS 10.14.6)  TLSv1.3   TLS_CHACHA20_POLY1305_SHA256      253 bit ECDH (X25519)
 Safari 15.4 (macOS 12.3.1)   TLSv1.3   TLS_AES_128_GCM_SHA256            253 bit ECDH (X25519)
 Java 7u25                    No connection
 Java 8u161                   TLSv1.2   ECDHE-RSA-AES128-GCM-SHA256       256 bit ECDH (P-256)
 Java 11.0.2 (OpenJDK)        TLSv1.3   TLS_AES_128_GCM_SHA256            256 bit ECDH (P-256)
 Java 17.0.3 (OpenJDK)        TLSv1.3   TLS_AES_128_GCM_SHA256            253 bit ECDH (X25519)
 go 1.17.8                    TLSv1.3   TLS_AES_128_GCM_SHA256            253 bit ECDH (X25519)
 LibreSSL 2.8.3 (Apple)       TLSv1.2   ECDHE-RSA-CHACHA20-POLY1305       253 bit ECDH (X25519)
 OpenSSL 1.0.2e               TLSv1.2   ECDHE-RSA-AES128-GCM-SHA256       256 bit ECDH (P-256)
 OpenSSL 1.1.0l (Debian)      TLSv1.2   ECDHE-RSA-AES128-GCM-SHA256       253 bit ECDH (X25519)
 OpenSSL 1.1.1d (Debian)      TLSv1.3   TLS_AES_128_GCM_SHA256            253 bit ECDH (X25519)
 OpenSSL 3.0.3 (git)          TLSv1.3   TLS_AES_128_GCM_SHA256            253 bit ECDH (X25519)
 Apple Mail (16.0)            TLSv1.2   ECDHE-RSA-AES128-GCM-SHA256       256 bit ECDH (P-256)
 Thunderbird (91.9)           TLSv1.3   TLS_AES_128_GCM_SHA256            253 bit ECDH (X25519)


 Rating (experimental)

 Rating specs (not complete)  SSL Labs's 'SSL Server Rating Guide' (version 2009q from 2020-01-30)
 Specification documentation  https://github.com/ssllabs/research/wiki/SSL-Server-Rating-Guide
 Protocol Support (weighted)  0 (0)
 Key Exchange     (weighted)  0 (0)
 Cipher Strength  (weighted)  0 (0)
 Final Score                  0
 Overall Grade                T
 Grade cap reasons            Grade capped to T. Issues with the chain of trust (self signed CA in chain)
                              Grade capped to M. Domain name mismatch
                              Grade capped to A. HSTS is not offered

 Done 2023-08-11 08:28:36 [  38s] -->> 127.0.0.1:9001 (localhost) <<--
```
/cc @sergiordlr
